train_gpt2.cu

```diff
@@ -733,8 +741,40 @@ void gpt2_backward(GPT2 *model, int* inputs, bool last_step) {
     if (model->grads_memory == NULL) {
         NvtxRange rng("InitGrads");
         // allocate buffers for weight gradients
-        printf0("allocating %d MiB for parameter gradients\n", (int)round(model->num_parameters * sizeof(floatX) / (1024 * 1024)));
-        model->grads_memory = malloc_and_point_parameters(&model->grads, model->param_elements, model->param_sizeof);
+        if(multi_gpu_config.zero_stage == 2) {
+            // Allocate parameter buffers for the current layers active "wave" of computation
+            size_t param_elements[NUM_PARAMETER_TENSORS];
+            size_t param_sizeof[NUM_PARAMETER_TENSORS];
+            GPT2Config wave_config = model->config;
+            // allocate space for two layers, so we can do double-buffering
+            wave_config.num_layers = 2;
+            fill_in_parameter_sizes(param_elements, param_sizeof, wave_config);
+            size_t alloc_bytes = 0;
+            for(int i = 0; i < NUM_PARAMETER_TENSORS; ++i) {
+                alloc_bytes += param_sizeof[i] * param_elements[i];
+            }
+            printf0("allocating %d MiB for ZeRO-2 active gradients\n",
+                    (int) round(alloc_bytes / (1024 * 1024)));
+            model->grads_memory = malloc_and_point_parameters(&model->grads, param_elements, param_sizeof);
+            model->grads_bytes = alloc_bytes;
+
+            alloc_bytes = 0;
+            fill_in_parameter_sizes(param_elements, param_sizeof, model->config);
+            for(int i = 0; i < NUM_PARAMETER_TENSORS; ++i) {
+                param_elements[i] /= multi_gpu_config.num_processes;
```

Comment:
Will reread the paper so my knowledge is fresh but i thought we'd be dividing by the size of the node here and not by the total number of processes